src/v/archival/tests/async_data_uploader_test.cc

```diff
@@ -0,0 +1,967 @@
+    };
+
+    /// Load arbitrary offset range using segment_upload class
+    std::optional<size_limited_upl_result>
+    read_offset_range(size_limited_offset_range range) {
+        vlog(
+          test_log.info,
+          "Query range: {}, max size {}, min size {}",
+          range.base,
+          range.min_size,
+          range.max_size);
+
+        auto partition = get_test_partition();
+        const auto& offsets = partition->log()->offsets();
+
+        vlog(
+          test_log.info,
+          "Partition offset range: {}-{}",
+          offsets.start_offset,
+          offsets.committed_offset);
+
+        for (const auto& s : get_partition_log()->segments()) {
+            vlog(
+              test_log.info,
+              "Segment {}, size {}, offsets {}",
+              s.get()->filename(),
+              s.get()->file_size(),
+              s.get()->offsets());
+        }
+
+        iobuf actual;
+        auto out_s = make_iobuf_ref_output_stream(actual);
+
+        auto upl_res = segment_upload::make_segment_upload(
+                         get_test_partition(),
+                         range,
+                         0x1000,
+                         ss::default_scheduling_group(),
+                         model::time_from_now(100ms))
+                         .get();
+
+        if (upl_res.has_failure()) {
+            vlog(
+              test_log.info,
+              "Failed to initialize segment_upload {}",
+              upl_res.error().message());
+            return std::nullopt;
+        }
+        auto actual_offset_range = upl_res.value()->get_meta().offsets;
+        auto meta_size_bytes = upl_res.value()->get_meta().size_bytes;
+
+        // Check metadata first
+        BOOST_REQUIRE_EQUAL(range.base, actual_offset_range.base);
+
+        auto upl_size = upl_res.value()->get_size_bytes();
+        auto meta = upl_res.value()->get_meta();
+        auto inp_s = std::move(*upl_res.value()).detach_stream().get();
+        auto closer = ss::defer([&] {
+            out_s.close().get();
+            inp_s.close().get();
+        });
+
+        ss::copy(inp_s, out_s).get();
+
+        vlog(
+          test_log.info,
+          "Actual size: {}, \n{}",
+          actual.size_bytes(),
+          actual.hexdump(1024));
+
+        BOOST_REQUIRE_GT(actual.size_bytes(), 0);
+
+        // Check size computation
+        vlog(
+          test_log.info,
+          "Computed size: {}, actual size: {}",
+          upl_size,
+          actual.size_bytes());
+        BOOST_REQUIRE_EQUAL(actual.size_bytes(), upl_size);
+        BOOST_REQUIRE_EQUAL(actual.size_bytes(), meta_size_bytes);
+
+        return size_limited_upl_result{
+          .payload = std::move(actual),
+          .range = actual_offset_range,
+          .meta = std::move(meta),
+        };
+    }
+
+    /// Load arbitrary offset range using segment_upload class
+    std::optional<iobuf> read_offset_range(inclusive_offset_range range) {
```

Comment:
nit: may be more quickly understood as `stream_upload_candidate_to_buf` or something, so it's clear at call sites this is referring to would-be-uploaded data